Logging/BaseLogHandler.cs

```diff
@@ -0,0 +1,148 @@
+        /// Write debug message to log
+        /// </summary>
+        /// <param name="text">The debug text to log</param>
+        public void Debug(string text)
+        {
+            var log = new LogEntry(text, DateTime.UtcNow, LogType.Debug);
+            _logsQueue.Enqueue(log);
+            _messageEvent.Set();
+        }
+
+        /// <summary>
+        /// Write debug message to log
+        /// </summary>
+        /// <param name="text">The trace text to log</param>
+        public void Trace(string text)
+        {
+            var log = new LogEntry(text, DateTime.UtcNow, LogType.Trace);
+            _logsQueue.Enqueue(log);
+            _messageEvent.Set();
+        }
+
+        /// <summary>
+        /// Thread function in charge of dispersing log messages to abstract write methods
+        /// </summary>
+        private void WriteThread()
+        {
+            LogEntry entry;
+
+            while (!_cancellationTokenSource.IsCancellationRequested && _messageEvent.WaitOne())
+            {   
+                try
+                {
+                    while (_logsQueue.TryDequeue(out entry))
+                    {
+                        switch(entry.MessageType)
+                        {
+                            case LogType.Trace:
+                                WriteTrace(entry.Message);
+                                break;
+                            case LogType.Debug:
+                                WriteDebug(entry.Message);
+                                break;
+                            case LogType.Error:
+                                WriteError(entry.Message);
+                                break;
+                        }
+                    }
+                } catch (Exception exception)
+                {
+                    Console.WriteLine("Logging thread has failed: " + exception);
+                    Thread.Sleep(10);
+                }
+            }
+        }
+
+        /// <summary>
+        /// Function to write a trace statement
+        /// </summary>
+        protected abstract void WriteTrace(string message);
+
+        /// <summary>
+        /// Function to write a debug statement
+        /// </summary>
+        protected abstract void WriteDebug(string message);
+
+        /// <summary>
+        /// Function to write a error statement
+        /// </summary>
+        protected abstract void WriteError(string message);
+
+        /// <summary>
+        /// Performs application-defined tasks associated with freeing, releasing, or resetting unmanaged resources.
+        /// </summary>
+        /// <filterpriority>2</filterpriority>
+        public virtual void Dispose()
+        {
+            try
+            {   
+                _cancellationTokenSource.Cancel();
+                _messageEvent.Set();
+                if (!_logThread.Join(1000))
+                {
+                    Console.WriteLine($"Timeout waiting for '{_logThread.Name}' thread to stop");
```

Comment:
I think in places of the code we re-route the console, so this may not be safe unless we first grab the reference in a static constructor and save it (`Console.Out`)

Reply:
True! I think we could actually directly call `WriteError` here, `ConsoleLogHandler` and `FileLogHandler` will work correctly, file handler is disposing after the base